incubator/hnc/internal/reconcilers/crd_removeversion.go

```diff
@@ -0,0 +1,106 @@
+package reconcilers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	apiextension "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// RemoveObsoleteCRDVersionReconciler tries to remove obsolete version from CRD.
+type RemoveObsoleteCRDVersionReconciler struct {
+	Client *apiextension.Clientset
+	Log    logr.Logger
+
+	// ObsoleteVersion is what we want to remove from the CRD status.storedVersions.
+	ObsoleteVersion string
+
+	// CRDNames is a set of the CRD names that we want to remove the obsolete
+	// version from its status.storedVersions.
+	CRDNames nameSet
+}
+
+// nameSet keeps track of a group of unique CRD names.
+type nameSet map[string]bool
+
+// Reconcile is the entry point to the reconciler.
+func (r *RemoveObsoleteCRDVersionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	// Early exit if the CRD is not the ones we want to remove the version.
+	if !r.CRDNames[req.Name] {
+		return ctrl.Result{}, nil
+	}
+
+	nm := req.Name
+	msg := fmt.Sprintf("Reconciling CRD %s to remove %s from status.storedVersions", nm, r.ObsoleteVersion)
```

Comment:
Better to use structured logging (here and elsewhere): `r.Log.Info("Ensuring obsolete storedVersions are removed", "name", nm, "oldVersion", r.ObsoleteVersion)`

Or even better:
```
log := r.Log.WithValues("name", nm, "oldVersion", r.ObsoleteVersion)
...
log.Info("Reconciling")
```

This ensures that the name is included if you call `log.Error`, which it's currently missing.

Reply:
Thanks for the suggestion! Yes, `.WithValues()` is much better, updated!